rct229/rules/section5/section5rule17.py

```diff
@@ -0,0 +1,158 @@
+                    surface_p_construction = surface_p["construction"]
+
+                if surface_b_type in [OST.ABOVE_GRADE_WALL, OST.FLOOR, OST.ROOF]:
+                    return {
+                        "id": surface_b["id"],
+                        "baseline_surface_type": surface_b_type,
+                        "proposed_surface_type": surface_p_type,
+                        "baseline_surface_u_factor": surface_b_construction.get("u_factor"),
+                        "proposed_surface_u_factor": surface_p_construction if surface_p_construction is None else
+                        surface_p_construction.get("u_factor")
+                    }
+                elif surface_b_type in [OST.UNHEATED_SOG, OST.HEATED_SOG]:
+                    return {
+                        "id": surface_b["id"],
+                        "baseline_surface_type": surface_b_type,
+                        "proposed_surface_type": surface_p_type,
+                        "baseline_surface_f_factor": surface_b_construction.get("f_factor"),
+                        "proposed_surface_f_factor": surface_p_construction if surface_p_construction is None else
+                        surface_p_construction.get("f_factor")
+                    }
+                elif surface_b_type == OST.BELOW_GRADE_WALL:
+                    return {
+                        "id": surface_b["id"],
+                        "baseline_surface_type": surface_b_type,
+                        "proposed_surface_type": surface_p_type,
+                        "baseline_surface_c_factor": surface_b_construction.get("c_factor"),
+                        "proposed_surface_c_factor": surface_p_construction if surface_p_construction is None else
+                        surface_p_construction.get("c_factor")
+                    }
+                else:
+                    # Will never reach this line
+                    # The OST defaults all unidentifiable surfaces to above wall grade
+                    # Serve code completeness
+                    return {
+                        "id": surface_b["id"],
+                        "baseline_surface_type": None,
+                        "proposed_surface_type": None,
+                        "baseline_surface_c_factor": None,
+                        "proposed_surface_c_factor": None
+                    }
+
+            def rule_check(self, context, calc_vals, data=None):
+                baseline_surface_type = calc_vals["baseline_surface_type"]
+                proposed_surface_type = calc_vals["proposed_surface_type"]
+                # Check 1. surface type needs to be matched
+                if proposed_surface_type is None or baseline_surface_type != proposed_surface_type:
+                    return False
+
+                if baseline_surface_type in [OST.ABOVE_GRADE_WALL, OST.FLOOR, OST.ROOF]:
+                    return calc_vals["proposed_surface_u_factor"] is not None or \
```

Comment:
I'm not following this logic. This seems to pass merely by having calc_vals["proposed_surface_u_factor"] not be None. Perhaps, you mean to use `and` instead of `or`.

Secondly, if we make u_factor required in this case (see above comment) then we don't need to handle these None cases here.

Thirdly, since we already have the baseline_surface_type cases logic in get_calc_vals(), here we could simplify to:
```
def rule_check(self, context, calc_vals, data=None):
    # Note: The factors that do not apply for the given baseline_surface_type are all set to None in get_calc_vals, so 
    the non-applicable factors are automatically equal here.

    return (calc_vals.get("baseline_surface_u_factor") == calc_vals.get("proposed_surface_u_factor")) and
        (calc_vals.get("baseline_surface_f_factor") == calc_vals.get("proposed_surface_f_factor")) and 
        (calc_vals.get("baseline_surface_c_factor") == calc_vals.get("proposed_surface_c_factor"))
```

Reply:
1. Yes, it supposes to be `and`
2. We cannot make u_factor as required since some constructions may only have c_factor or f_factor.
3. This works except for both None situation - follow the above comment, maybe we need to clarify the outcome for both = None situation. Fail or Undermined (Undetermined will be much simpler...)